### ISSUE TITLE
Couple of samples in readme

### ISSUE THREAD

**Issue by ogencoglu** [MinishLab/semhash#31]
Can you include couple of samples that are deemed to be deduplicates in your readme?

Let's say how would two samples such as:

`When photon energies are as high as the electron rest energy of 511 keV, yet another process, Compton scattering, may occur. Above twice this energy, at 1.022 MeV, pair production is also more likely. Compton scattering and pair production are examples of two other competing mechanisms. Even if the photoelectric effect is the favoured reaction for a particular interaction of a single photon with a bound electron, the result is also subject to quantum statistics and is not guaranteed. The probability of the photoelectric effect occurring is measured by the cross section of the interaction, σ. This has been found to be a function of the atomic number of the target atom and photon energy.`

`When photon energies are as high as the electron rest energy of 511 keV, yet another process, Compton scattering, may occur. Above twice this energy, at 1.032 MeV, pair production is also more likely. Compton scattering and pair production are examples of two other competing mechanisms. Even if the photoelectric effect is the favoured reaction for a particular interaction of a single photon with a bound electron, the result is also subject to quantum statistics and is not guaranteed. The probability of the photoelectric effect occurring is measured by the cross section of the interaction, σ. This has been found to be a function of the atomic number of the target atom and photon energy.`

be treated?

They are exactly the same except the second one has the value `1.032` instead of `1.022`. These sort of small factual differences are very important for hallucination checks and factuality evaluations.

**Comment by stephantul** [comment]
Hey, @ogencoglu good idea! We'll see how we can formalize this.

For what it's worth, these get a similarity of 0.998 using the default model, [potion-base-8m](https://huggingface.co/minishlab/potion-base-8M).

Out of curiosity: would these samples having such a high similarity be good or bad for your use-case?

**Comment by ogencoglu** [comment]
Thanks a lot for the swift reply.

In most factuality, hallucinations etc cases those two samples are completely different. That's why pure vector similarity or things of that nature doesn't work in these contexts and people are implementing hybrid stuff (combining old-school bm25s etc).

**Comment by sky-2002** [comment]
@ogencoglu interesting point, I have also observed that embeddings don't work in case of exact negations (like `I am a badminton player` and `i am not a badminton player` will have high similarity despite of being opposites). 

Also the factual correctness point you made is so valid, I am literally facing this issue where the LLM goofs up numbers.

**Comment by ogencoglu** [comment]
@sky-2002 yep.

Factuality, precision and things of that sort are more important than aggregated semantic similarity stuff.

**Comment by Pringled** [comment]
Hi, I think this is a slightly different than what SemHash is intended to do. Spotting or verifying hallucinations is an interesting problem in itself, but indeed won't work with pure vector similarity, just like the negation example. SemHash is intended to clean up (pre)training datasets and remove train/test leakage, in a fast, scalable, and lightweight manner.

**Comment by ogencoglu** [comment]
@Pringled  Nobody is trying to use SemHash or vector similarity for halu detection. The point is SemHash is not useful for curating the pretraining data for the most value-creating LLM pipelines, i.e., the ones that require precision, factuality and reliability. That's where most of the GenAI value is.

**Comment by Pringled** [comment]
@ogencoglu I don't think this is true. There have been numerous studies that show that deduplicating pre-training data makes LLMs less likely to output exact samples from their pre-training data (e.g. https://arxiv.org/pdf/2107.06499). There is an interesting section in the HuggingFace blogpost on creating FineWeb that covers the effects of deduplication for their dataset (see https://huggingface.co/spaces/HuggingFaceFW/blogpost-fineweb-v1). Most of these rely on MinHash, or similar algorithms, which would also flag negations as duplicates due to their high string similarity.

**Comment by ogencoglu** [comment]
> deduplicating pre-training data makes LLMs less likely to output exact samples from their pre-training data

So? You don't need a paper to show that. Training means minimizing the loss on training samples, yes. How is that related to the topic?

If I have the following sentence verbatim 1000 times in a corpus, I want to of course remove the duplicates:

`When photon energies are as high as the electron rest energy of 511 keV, yet another process, Compton scattering, may occur. Above twice this energy, at 1.022 MeV, pair production is also more likely. Compton scattering and pair production are examples of two other competing mechanisms. Even if the photoelectric effect is the favoured reaction for a particular interaction of a single photon with a bound electron, the result is also subject to quantum statistics and is not guaranteed. The probability of the photoelectric effect occurring is measured by the cross section of the interaction, σ. This has been found to be a function of the atomic number of the target atom and photon energy.`

But I absolutely want to keep the variants where the factual values were different (let's say `511` vs `611` etc.). There are many use cases where I wouldn't care if the non-factual part is more likely to be memorized as long as the log-probs of the factual parts and named entities are returning low due to different views in the training set.

I am not saying this needs to be solved by this library. I am just saying.

**Comment by Pringled** [comment]
You specifically said "The point is SemHash is not useful for curating the pretraining data for the most value-creating LLM pipelines". My point is that this statement is not true, and there's evidence to support this. I'm not trying to argue that SemHash (or MinHash, or any existing deduplication technique) solves your specific problem. Data deduplication will inevitably lead to removing samples that you most likely did not want to remove. In my opinion, this trade-off is worth it in most usecases where you're using an uncurated dataset.

**Comment by Pringled** [comment]
Closing this issue for now. Feel free to open a new issue in case of any other questions 👍
